File: apps/client-macos/Sources/TimeTrack/Activity/ActivitySampleStore.swift
import Foundation

/// PRD §7.5 — durable, file-backed JSON buffer for activity samples. Mirrors `ImageBufferStore`'s
/// durability model (atomic write-tmp-then-rename, `.tmp-` sweep on init, one file per record,
/// FIFO by filename-encoded millis) but stores JSON and drains in BATCHES (its own uploader/engine),
/// so it is separate from the one-POST-per-record `BufferStore`. Filename `<createdMillis>__<id>.json`
/// carries order + identity without reading the file.
protocol ActivitySampleBuffering {
    func enqueue(_ sample: ActivitySample)
    func take(limit: Int) -> [ActivitySample]
    func remove(ids: [String])
    func prune(olderThan maxAge: TimeInterval, maxCount: Int)
    func clear()
}

final class ActivitySampleStore: ActivitySampleBuffering {
    static let shared = ActivitySampleStore(directory: ActivitySampleStore.defaultDirectory())

    private let directory: URL
    private let clock: () -> Date

    init(directory: URL, clock: @escaping () -> Date = Date.init) {
        self.directory = directory
        self.clock = clock
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        sweepTemporaries()
    }

    /// ~/Library/Application Support/<container>/activity/ — per-install; see `AppInstall`.
    static func defaultDirectory() -> URL {
        AppInstall.supportDirectory("activity")
    }

    func enqueue(_ sample: ActivitySample) {
        guard let payload = try? JSONEncoder().encode(sample) else { return }
        let millis = Int64(clock().timeIntervalSince1970 * 1000)
        let dst = directory.appendingPathComponent("\(millis)__\(sample.id).json")
        let tmp = directory.appendingPathComponent(".tmp-\(sample.id)")
        do {
            try payload.write(to: tmp, options: .atomic)
            try? FileManager.default.removeItem(at: dst)
            try FileManager.default.moveItem(at: tmp, to: dst)
        } catch {
            try? FileManager.default.removeItem(at: tmp)
        }
    }

    func take(limit: Int) -> [ActivitySample] {
        allRecords().prefix(limit).compactMap { rec in
            guard let data = try? Data(contentsOf: rec.url) else { return nil }
            return try? JSONDecoder().decode(ActivitySample.self, from: data)
        }
    }

    func remove(ids: [String]) {
        let set = Set(ids)
        for rec in allRecords() where set.contains(rec.id) {
            try? FileManager.default.removeItem(at: rec.url)
        }
    }

    /// Age-bound (drop older than `maxAge`) then count-bound (trim oldest beyond `maxCount`).
    func prune(olderThan maxAge: TimeInterval, maxCount: Int) {
        let cutoff = Int64(clock().addingTimeInterval(-maxAge).timeIntervalSince1970 * 1000)
        var records = allRecords()
        for rec in records where rec.createdMillis < cutoff {
            try? FileManager.default.removeItem(at: rec.url)
        }
        records = records.filter { $0.createdMillis >= cutoff }
        if records.count > maxCount {
            for rec in records.prefix(records.count - maxCount) {
                try? FileManager.default.removeItem(at: rec.url)
            }
        }
    }

    func clear() {
        for rec in allRecords() { try? FileManager.default.removeItem(at: rec.url) }
    }

    // MARK: - internals

    private struct Record { let url: URL; let createdMillis: Int64; let id: String }

    private func contents() -> [URL] {
        (try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)) ?? []
    }

    private func parse(_ url: URL) -> Record? {
        let name = url.lastPathComponent
        guard name.hasSuffix(".json") else { return nil }
        let parts = String(name.dropLast(5)).components(separatedBy: "__")
        guard parts.count == 2, let millis = Int64(parts[0]) else { return nil }
        return Record(url: url, createdMillis: millis, id: parts[1])
    }

    private func allRecords() -> [Record] {
        contents().compactMap(parse).sorted { ($0.createdMillis, $0.id) < ($1.createdMillis, $1.id) }
    }

    private func sweepTemporaries() {
        for url in contents() where url.lastPathComponent.hasPrefix(".tmp-") {
            try? FileManager.default.removeItem(at: url)
        }
    }
}

File: apps/client-windows/src/NiftyTimer/Storage/ActivitySampleStore.cs
using System.Globalization;
using System.Text.Json;
using NiftyTimer.Sync;

namespace NiftyTimer.Storage;

/// <summary>The activity-buffer seam. Tests substitute a spy.</summary>
public interface IActivitySampleBuffer
{
    void Enqueue(ActivitySamplePayload sample);

    IReadOnlyList<ActivitySamplePayload> Take(int limit);

    void Remove(IEnumerable<string> ids);

    void Prune(TimeSpan maxAge, int maxCount);

    void Clear();

    int PendingCount();
}

/// <summary>
/// PRD §7.5 — durable, file-backed buffer for activity samples.
///
/// Shares <see cref="BufferStore"/>'s durability model (one file per record, write to
/// <c>.tmp-&lt;id&gt;</c> then rename, sweep leftover temporaries on construction, FIFO by a
/// filename-encoded millisecond stamp) but is a SEPARATE store because it drains in BATCHES
/// through its own uploader and engine, where <see cref="BufferStore"/> posts one record per
/// request. Filename <c>&lt;createdMillis&gt;__&lt;id&gt;.json</c> carries order and identity
/// without opening the file.
///
/// Not a capture path: this holds records that have already been captured, so it takes no gate
/// and lives in <c>NiftyTimer.Storage</c> rather than a capture namespace.
/// </summary>
public sealed class ActivitySampleStore : IActivitySampleBuffer
{
    private readonly string _directory;
    private readonly Func<DateTimeOffset> _clock;

    public ActivitySampleStore(string directory, Func<DateTimeOffset>? clock = null)
    {
        _directory = directory;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
        Directory.CreateDirectory(directory);
        SweepTemporaries();
    }

    public void Enqueue(ActivitySamplePayload sample)
    {
        byte[] payload;
        try
        {
            payload = JsonSerializer.SerializeToUtf8Bytes(sample, TimeEntryPayload.JsonOptions);
        }
        catch (NotSupportedException)
        {
            return;
        }

        var millis = _clock().ToUnixTimeMilliseconds();
        var dst = Path.Combine(_directory, $"{millis}__{sample.Id}.json");
        var tmp = Path.Combine(_directory, $".tmp-{sample.Id}");
        try
        {
            File.WriteAllBytes(tmp, payload);
            File.Move(tmp, dst, overwrite: true);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            TryDelete(tmp);
        }
    }

    /// <summary>FIFO (oldest first), up to <paramref name="limit"/>.</summary>
    public IReadOnlyList<ActivitySamplePayload> Take(int limit)
    {
        var result = new List<ActivitySamplePayload>();
        foreach (var rec in AllRecords())
        {
            if (result.Count >= limit)
            {
                break;
            }

            try
            {
                if (JsonSerializer.Deserialize<ActivitySamplePayload>(File.ReadAllBytes(rec.Path)) is { } sample)
                {
                    result.Add(sample);
                }
            }
            catch (Exception e) when (e is IOException or JsonException)
            {
                // Unreadable or corrupt mid-drain. Skipped, not deleted: Prune bounds the buffer,
                // and deleting on a read error would discard a record that a retry might recover.
            }
        }

        return result;
    }

    public void Remove(IEnumerable<string> ids)
    {
        var set = new HashSet<string>(ids, StringComparer.Ordinal);
        if (set.Count == 0)
        {
            return;
        }

        foreach (var rec in AllRecords())
        {
            if (set.Contains(rec.Id))
            {
                TryDelete(rec.Path);
            }
        }
    }

    /// <summary>Age-bound first (drop anything older than <paramref name="maxAge"/>), then
    /// count-bound (trim the oldest beyond <paramref name="maxCount"/>).</summary>
    public void Prune(TimeSpan maxAge, int maxCount)
    {
        var cutoff = _clock().Subtract(maxAge).ToUnixTimeMilliseconds();
        var surviving = new List<Record>();
        foreach (var rec in AllRecords())
        {
            if (rec.CreatedAtMillis < cutoff)
            {
                TryDelete(rec.Path);
            }
            else
            {
                surviving.Add(rec);
            }
        }

        for (var i = 0; i < surviving.Count - maxCount; i++)
        {
            TryDelete(surviving[i].Path);
        }
    }

    public void Clear()
    {
        foreach (var rec in AllRecords())
        {
            TryDelete(rec.Path);
        }
    }

    public int PendingCount() => AllRecords().Count;

    private readonly record struct Record(string Path, long CreatedAtMillis, string Id);

    private string[] Contents()
    {
        try
        {
            return Directory.GetFiles(_directory);
        }
        catch (DirectoryNotFoundException)
        {
            return [];
        }
    }

    private static Record? Parse(string path)
    {
        var name = Path.GetFileName(path);
        if (!name.EndsWith(".json", StringComparison.Ordinal))
        {
            return null;
        }

        var parts = name[..^5].Split("__");
        if (parts.Length != 2 ||
            !long.TryParse(parts[0], NumberStyles.Integer, CultureInfo.InvariantCulture, out var millis))
        {
            return null;
        }

        return new Record(path, millis, parts[1]);
    }

    private List<Record> AllRecords()
    {
        var records = new List<Record>();
        foreach (var path in Contents())
        {
            if (Parse(path) is { } rec)
            {
                records.Add(rec);
            }
        }

        records.Sort(static (a, b) =>
        {
            var byTime = a.CreatedAtMillis.CompareTo(b.CreatedAtMillis);
            return byTime != 0 ? byTime : string.CompareOrdinal(a.Id, b.Id);
        });
        return records;
    }

    private void SweepTemporaries()
    {
        foreach (var path in Contents())
        {
            if (Path.GetFileName(path).StartsWith(".tmp-", StringComparison.Ordinal))
            {
                TryDelete(path);
            }
        }
    }

    private static void TryDelete(string path)
    {
        try
        {
            File.Delete(path);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            // Another pass got there first, or the file is briefly locked. The next cycle retries;
            // a failed delete must never abort a drain.
        }
    }
}
